### PR TITLE
Enable Thumbless CropDuster 

### PR DESCRIPTION
**Pull request by obswork**
Adds a settings flag that can be used to toggle whether or not thumb files are produced. They're still saved to the db, which allows them to be used as bounding box hints for services such as e.g. Thumbor. JS changes ensure that a preview image is still rendered in the admin. Along with the original image, the preview is the only other image file saved to the storage backend when the settings flag is enabled. 

(will be released as 4.12.9)

refs: FUN-2528

### CODE REVIEW

File: cropduster/tests/test_models.py [hunk 1]
```diff
@@ -34,6 +39,7 @@ def test_original_image(self):
         self.assertIs(article.lead_image.sizes, Article.LEAD_IMAGE_SIZES)
 
     def test_generate_thumbs(self):
+        reload(cropduster_settings)
```
Comment (fdintino) [inline cropduster/tests/test_models.py:42]: I know this isn't up for code review yet, but I think what you actually want to use here is the [`setting_changed`](https://docs.djangoproject.com/en/3.0/ref/signals/#setting-changed) signal. Here is an example of its usage from django-pipeline:

https://github.com/jazzband/django-pipeline/blob/ce19344ef414770753b2d8ae310e6d778dc3599e/pipeline/conf.py#L121-L124

```python
@receiver(setting_changed)
def reload_settings(**kwargs):
    if kwargs['setting'] == 'PIPELINE':
        settings.update(kwargs['value'])
```

In truth, it's an anti-pattern to grab these settings and set them in a module, but short of a refactor of how settings are referenced internally in the app, the simplest option is to put a receiver like this in cropduster/settings.py

Comment (obswork) [reply cropduster/tests/test_models.py:42]: Cool, i'll give that a try 👍

Comment (fdintino) [reply cropduster/tests/test_models.py:42]: With the caveat that instead of a settings dict you would have `global CROPDUSTER_CREATE_THUMBS` etc etc

File: cropduster/tests/helpers.py [hunk 2]
```diff
@@ -39,26 +39,6 @@ def setUp(self):
         self.TEST_IMG_DIR = ORIG_IMG_PATH
         self.TEST_IMG_DIR_RELATIVE = os.path.join(random, 'data')
 
-    def assertImageColorEqual(self, element, image):
```
Comment (fdintino) [inline cropduster/tests/helpers.py:42]: Why was this deleted?

Comment (fdintino) [reply cropduster/tests/helpers.py:42]: To explain why it's here: this was written to catch a common class of bug, which is that after uploading and cropping, the image preview might not appear.

Comment (fdintino) [reply cropduster/tests/helpers.py:42]: I just re-enabled this check and ran it locally, and I think this caught a real bug in this PR. In addition to this failing with `AssertionError: Colors differ: #ffffff != #000000` I see in the console the error `Failed to load resource: the server responded with a status of 404 (Not Found) _preview.png`

Comment (fdintino) [reply cropduster/tests/helpers.py:42]: Sorry to be carrying on a conversation with myself, but I was curious about this and looked into it more: this particular bug is not a real failure, though I'm curious whether  It seems that the issue here is that `_preview.ext` doesn't get created by `generate_thumbs`. The quickest way to fix would be to add some code to `create_unique_image` in the cropduster test helpers that also saves a file to that path (it doesn't really matter what the dimensions are, it can be the same as the original for the purpose of these checks).

File: cropduster/static/cropduster/js/cropduster.js [hunk 3]
```diff
@@ -372,6 +372,20 @@ window.CropDuster = {};
                     $thumb.html($thumb.html() + $.render.cropdusterImage(thumbData[size.name]));
                 }
             });
+            if ($thumb) {
```
Comment (fdintino) [inline cropduster/static/cropduster/js/cropduster.js:375]: For jquery truthiness, you'll want `if ($thumb.length)` here (an empty jQuery object still evaluates as truthy). Out of curiosity: what problem is this fixing?

File: cropduster/tests/helpers.py (hunk 2 above)
Comment (fdintino) [reply cropduster/tests/helpers.py:42]: It didn't cause an issue before because it would use the first available thumb, instead of `_preview.ext`. When the change was made in this PR to cropduster.js to use _preview.ext instead, that introduced the test failure.

File: cropduster/tests/test_models.py (hunk 1 above)
Comment (obswork) [reply cropduster/tests/test_models.py]: Gave this a go, more-or-less worked :+1: But (I think) because of the way we actually read this (and other settings) off `cropduster_settings`, the global method doesn't work. Patching it directly like so seemed to do the trick, though-
```
 @receiver(setting_changed)                                                                                            
 def reload_settings(**kwargs):                                                                                        
     if kwargs['setting'] == 'CROPDUSTER_CREATE_THUMBS':                                                               
         from cropduster import settings as cropduster_settings                                                        
         cropduster_settings.CROPDUSTER_CREATE_THUMBS = kwargs['value']    
```

File: cropduster/tests/helpers.py (hunk 2 above)
Comment (obswork) [reply cropduster/tests/helpers.py]: Reinstated the test and added shtuff to make a preview image in `create_unique_thumbs` as suggested. Got rid of the test initially because I knew the failure was spurious (based on manual upload tests) and..I was too lazy to dig much further :sweat_smile:

File: cropduster/static/cropduster/js/cropduster.js (hunk 3 above)
Comment (obswork) [reply cropduster/static/cropduster/js/cropduster.js]: Eh, it was an unnecessary guard class. Didn't solve a problem and is now no longer not solving a problem :smile:

File: cropduster/static/cropduster/js/cropduster.js [hunk 4]
```diff
@@ -372,6 +372,18 @@ window.CropDuster = {};
                     $thumb.html($thumb.html() + $.render.cropdusterImage(thumbData[size.name]));
                 }
             });
+            var url = [CropDuster.mediaUrl, path, '_preview' + ext].join('/');
```
Comment (fdintino) [inline cropduster/static/cropduster/js/cropduster.js:375]: The code in this file is a nightmare, so it's an easy mistake but: I think this introduces a bug. This uses the definition of `url` in the loop above. But notice that we don't actually pass that value to the template renderer any more. If the media storage is something other than django's FileSystemStorage, we can't reliably construct the url in this way; we have to get it from the server.

Comment (fdintino) [reply cropduster/static/cropduster/js/cropduster.js:375]: Confirmed. This assertion is removed in this PR, but if put back in the S3 test fails:

```patch
diff --git a/cropduster/tests/test_admin.py b/cropduster/tests/test_admin.py
index 7d7a8ed..1b3c158 100644
--- a/cropduster/tests/test_admin.py
+++ b/cropduster/tests/test_admin.py
@@ -148,7 +148,6 @@ def test_changeform_single_image(self):
 
         preview_image_el = self.selenium.find_element_by_css_selector('#headshot-group .cropduster-image span')
         src_image_path = os.path.join(self.TEST_IMG_DIR, 'img.png')
-        self.assertImageColorEqual(preview_image_el, src_image_path)
 
         elem = self.selenium.find_element_by_id('id_name')
         elem.clear()
```

Comment (fdintino) [reply cropduster/static/cropduster/js/cropduster.js:375]: You'll need to add a data attribute here with the preview image's url:

https://github.com/theatlantic/django-cropduster/blob/bc251f421ea2b8e8a7370ad55d221394db65e344/cropduster/templates/cropduster/custom_field.html#L6-L12

You can pass context data to that template here:

https://github.com/theatlantic/django-cropduster/blob/bc251f421ea2b8e8a7370ad55d221394db65e344/cropduster/forms.py#L31-L46

Comment (obswork) [reply cropduster/static/cropduster/js/cropduster.js:375]: A-ha, very helpful. Cheers :+1: I'll give that a go.

Comment (fdintino) [reply cropduster/static/cropduster/js/cropduster.js:375]: That will ensure that the preview image shows up when the admin changeview is loaded. In order to also update that image after a user has finished uploading and cropping an image, you'll need to add the preview url to the crop view's response:

https://github.com/theatlantic/django-cropduster/blob/178cb00a6a19660d40235e8ec30c65175c158f62/cropduster/views/__init__.py#L468-L472

which gets passed along in upload.js to `parent.CropDuster.complete`:

https://github.com/theatlantic/django-cropduster/blob/178cb00a6a19660d40235e8ec30c65175c158f62/cropduster/static/cropduster/js/upload.js#L703

which is defined here:

https://github.com/theatlantic/django-cropduster/blob/178cb00a6a19660d40235e8ec30c65175c158f62/cropduster/static/cropduster/js/cropduster.js#L235-L250

Where you'll need to do something like this before the call to `createThumbnails`:

```javascript
$("#id_" + prefix).data('previewUrl', data.preview_url);
```

File: cropduster/models.py [hunk 5]
```diff
@@ -538,9 +538,11 @@ def _save_thumb(self, size, image=None, thumb=None, ref_thumb=None, tmp=False, c
 
         thumb_crop = thumb.crop(image, size)
         thumb_path = self.get_image_path(size.name, tmp=tmp)
-        thumb_image = thumb_crop.create_image(thumb_path, width=thumb.width, height=thumb.height)
 
-        if StandaloneImage:
+        if cropduster_settings.CROPDUSTER_CREATE_THUMBS:
```
Comment (fdintino) [inline cropduster/models.py:542]: I think we'll need to exempt cropduster standalone images from this conditional. These are the images that can be uploaded via the ckeditor plugin. We render a single "thumb" for the image, and I think doing so is unavoidable. The alternative would be to somehow pass the thumbor url back to the ckeditor plugin, and that would require more changes in our image html processor that adds retina `srcset` attributes to images.

Comment (obswork) [reply cropduster/models.py:542]: Looks like the cropduster standalones are exempted by virtue of using a different method for their thumb creation- [_save_standalone_thumb](https://github.com/theatlantic/django-cropduster/blob/master/cropduster/models.py#L481).

File: cropduster/static/cropduster/js/cropduster.js (hunk 4 above)
Comment (obswork) [reply cropduster/static/cropduster/js/cropduster.js]: Done 👍
